ci/create_yum_repo/main.py

```diff
@@ -0,0 +1,135 @@
+        upload_task = {}
+
+        for filename in walk_directory(directory):
+            upload_task[executor.submit(upload_file, filename)] = filename
+
+        for task in futures.as_completed(upload_task):
+            try:
+                task.result()
+            except Exception as e:
+                logging.error(
+                    f"Exception {e} encountered while uploading file"
+                    f"{upload_task[task]}"
+                )
+                ret = 1
+    return ret
+
+
+def create_repo(repodir: str) -> int:
+    return utils.run_cmd(cmd=f"createrepo_c {repodir}")
+
+
+def test_repo(bucket: str, region: str, id: int) -> int:
+    with open("/etc/yum.repos.d/automotive.repo", "w") as repofile:
+        repofile.write(
+            Template(
+                open(
+                    os.path.join(
+                        (os.path.dirname(os.path.realpath(__file__))),
+                        "yumrepo.j2",
+                    ),
+                ).read()
+            ).render(bucket=bucket, region=region, id=id)
+        )
+    return utils.run_cmd(cmd="sudo dnf makecache --repo=automotive")
+
+
+def main() -> int:
+    ret0 = 0
+    repodir = "/var/lib/repos"
+    os.mkdir(f"{repodir}")
+
+    s = requests.Session()
+    s.mount("http://mirror.centos.org", HTTPAdapter(max_retries=5))
+
+    with open("./package_list/c8s-image-manifest.txt") as im:
+        for p in im.readlines():
+            repositories = ["BaseOS", "AppStream"]
+            arches = ["aarch64", "noarch"]
+            for repo in repositories:
+                for arch in arches:
+                    pkg = f"{p.strip()}.{arch}.rpm"
+                    r = s.get(
+                        url=(
+                            f"http://mirror.centos.org/centos/8-stream/{repo}/"
+                            f"aarch64/os/Packages/{pkg}"
+                        ),
+                        allow_redirects=True,
+                    )
+                    if r.status_code == 200:
+                        with open(f"{repodir}/{pkg}", "wb") as f:
+                            f.write(r.content)
+                        logging.debug(f"{pkg} downloaded")
+                        break  # pkg was found and was downloaded
+                else:  # pkg was not found in any of the arches
+                    continue  # continue to next repository.
+                break  # pkg was found and was downloaded
+            else:  # pkg was not found in any of the arches*repositories
+                logging.error(f"{p.strip()} was not found")
+                ret0 = 1
+
+    ret1 = create_repo(repodir)
+
+    ret2 = upload_directory(
+        directory=repodir,
+        prefix=os.getenv("GITHUB_RUN_ID"),
+    )
+
+    ret3 = test_repo(
+        bucket=os.getenv("AWS_BUCKET_REPOS"),
+        region=os.getenv("AWS_REGION"),
+        id=os.getenv("GITHUB_RUN_ID"),
+    )
+    with open("/etc/yum.repos.d/automotive.repo") as f:
+        logging.info(f.read())
```

Comment:
Can't we check the log lvl first, so that we wouldn't even open the file otherwise?

Reply:
one of the idea for log levels is to have filter post. We shouldn't log on condition

Reply:
what doesn't it mean - "to have filer post" ?

Reply:
typo , "filter". We shouldn't have conditional logs. This filtering should be done post by a log aggregator.

Reply:
ack